### ISSUE TITLE
Cache doesn't cull unreachable tasks after replacement

### ISSUE THREAD

**Issue by gjoseph92** [dask/dask#6228]
Using the `Cache` class, tasks which are no longer needed—because their results have been found in the cache—may still be run.

```python
In [1]: import time
In [2]: import dask
In [3]: import dask.cache

In [4]: @dask.delayed
   ...: def slow(x):
   ...:     print(f"slow {x}")
   ...:     time.sleep(2)
   ...:     return x
   ...:

In [5]: @dask.delayed
   ...: def fast_add(x, y):
   ...:     print(f"fast_add {x} {y}")
   ...:     return x + y
   ...:

In [6]: result = fast_add(slow(1), slow(2))

In [7]: cache = dask.cache.Cache(1e6)

In [8]: # as expected, the first run takes 2sec and all functions run
   ...: %%time
   ...: with cache:
   ...:     result.compute()
   ...:
slow 1
slow 2
fast_add 1 2
CPU times: user 6.66 ms, sys: 3.49 ms, total: 10.1 ms
Wall time: 2.01 s

In [9]: # also as expected, the second run is fast, since all results are in the cache---no functions run
   ...: %%time
   ...: with cache:
   ...:     result.compute()
   ...:
CPU times: user 422 µs, sys: 24 µs, total: 446 µs
Wall time: 442 µs

In [10]: cache.cache.data
Out[10]:
{'slow-3d3a6713-ce9f-49ea-9308-a765127fa68e': 1,
 'slow-670ba8b8-6ab1-4843-87d2-c5bdc7e0a730': 2,
 'fast_add-2432fe5a-e323-4bc5-b11d-7f357ac40f55': 3}

In [11]: # what if we had a cache miss on the upstream `slow` functions,
    ...: # but still a cache hit on the result we're actually trying to compute?
    ...: # we'd expect the slow functions to still not run, because with the result
    ...: # already in the cache, those upstream dependencies are essentially unreachable
    ...: # keys---nothing depends on them anymore.

In [11]: del cache.cache.data['slow-3d3a6713-ce9f-49ea-9308-a765127fa68e']
    ...: del cache.cache.data['slow-670ba8b8-6ab1-4843-87d2-c5bdc7e0a730']

In [12]: # to confirm: the only result key we need is still in the cache
    ...: result.__dask_keys__()                                                                                                                                                               
Out[12]: ['fast_add-2432fe5a-e323-4bc5-b11d-7f357ac40f55']

In [13]: cache.cache.data                                                                                                                                                                     
Out[13]: {'fast_add-2432fe5a-e323-4bc5-b11d-7f357ac40f55': 3}

In [14]: # yet the `slow` functions still run, even though they're unneeded. `fast_add` doesn't, because it's a cache hit.
    ...: %%time
    ...: with cache:
    ...:     result.compute()
    ...:
slow 1
slow 2
CPU times: user 3.44 ms, sys: 1.46 ms, total: 4.9 ms
Wall time: 2.01 s
```

The Cache class simply inserts cached values into the dask prior to computation:
https://github.com/dask/dask/blob/3b92efff2e779f59e95e05af9b8f371d56227d02/dask/cache.py#L47-L49

However, the implementation of `get_async` assumes the graph it's passed has already been optimized and culled. Since Cache is a Callback, and Callbacks operate post-optimization, the graph it produces won't ever get optimized. Therefore, the now-unreachable tasks in the graph are still computed.

There are a couple ways we could solve this:

1. Have `get_async` always cull the graph. IMO, the contract of `get_async` is just that it'll compute the `result` keys from the given `dsk`—_not_ that it'll execute every task in the graph. To me, it's just an unfortunate implementation detail that it schedules tasks to run which are not direct dependencies of the `result` keys.

    I wouldn't necessarily implement this by just calling `cull` in `get_asyc` (though it may not be the worst idea). Instead, I imagine that [`start_state_from_dask`](https://github.com/dask/dask/blob/3b92efff2e779f59e95e05af9b8f371d56227d02/dask/local.py#L139-L198) could be refactored in such a way that it walks the graph starting from the result keys, which would probably be about the same performance-wise, but have the property of culling the graph along the way.

2. The Callback interface doesn't seem suited to callbacks that modify the graph, because they're not told which keys are being requested. If `_start` were called with `(dsk, keys)` instead of just `(dsk)`, then the callback could just call `cull` itself inside `_start`. But without `keys`, it doesn't have enough information to do any optimizations.

    This would be a pretty simple change, but a breaking change to the Callback interface.

3. Perhaps callbacks really shouldn't be modifying the graph like this—should cache-replacement be considered an optimizer instead? That interface calls `optimize(dsk, keys)`, so it would work. `Cache.__enter__` could temporarily add a function to `config["optimizations"]` that replaces the keys in the graph with their cached values, then culls it.

    However, that won't work with the current [`collections_to_dsk`](https://github.com/dask/dask/blob/3b92efff2e779f59e95e05af9b8f371d56227d02/dask/base.py#L204-L231) implementation, which is where `config["optimizations"]` are applied: it applies `config["optimizations"]` _before_ any collection-specific optimizations. Since the cache uses the `_posttask` callback for cache puts, where the keys are in their fully-optimized form, the cache gets also need to operate on the fully-optimized graph. Basically, cache replacement always has to be the very last optimization pass, and there's no interface for that right now.

    Two options for this:
        a. Flip these sections:
        https://github.com/dask/dask/blob/e3f80df7fff7627082688b7cccff1e9131001446/dask/base.py#L213-L221

      so that `optimizations` are applied after the collection-specific optimizations. Easy, but possibly a breaking change for anyone using the `optimizations=` interface.

      EDIT: looks like this flip actually just happened in #6137?! Hope that doesn't break anything for anyone...

      b. Add another kwarg for `"post_optimizations"`?

For context, I found this because I wrote a custom caching callback, and referred to the `Cache` implementation while writing it. I recently found this bug in mine, and because of all the limitations I've described above, I can't find any way to work around it without making some change to dask.

Which direction (cull in `get_async`, pass keys to `_start`, or register in `config["optimizations"]`) do you think would be the best way to solve this?

**Comment by martindurant** [comment]
Thanks for posting this thoughtful discussion. 

From your listed options, it seems to me that "start_state_from_dask could be refactored" is the minimally-invasive one that achieves your goal and may indeed simplify the code. Have you tried to scope this yet or otherwise have thoughts on how complicated t would be? Otherwise, calling cull may be ok, not so sure on where the tradeoffs for that would be.

**Comment by mrocklin** [comment]
Yeah, nice writeup @gjoseph92 .  Given how easy it is to forget things, I suspect that you may now be the world expert on this topic, probably eclipsing @jcrist who wrote a lot of the callback machinery :).  One answer is that all of these seem like good optimizations (regardless of this specific application problem) if we do them well:

1.  Culling as a side effect of graph traversal sounds great, if it can be done without performance penalty
2.  Adding keys to the Callback interface sounds great, _if_ it can be done in a backwards compatible way.  Perhaps we ask if the start function has a `keys` parameter and if so, we pass in the keys.  See `dask.utils.has_keyword`
3.  We kind of want caching to happen in the middle of optimizations, before fusion but after culling. There has been some conversation about how to make the optimization pipeline more configurable. (cc @sofroniewn who ran into this with Napari)

**Comment by mrocklin** [comment]
Between these three options, number 2 seems like the easiest to me.  Number 3 would have the most positive impact, but is the least well defined.

**Comment by sofroniewn** [comment]
cc @tlambert03
